// jsreport-core/lib/worker/reporter.js
const { nanoid } = require('nanoid')
const ExtensionsManager = require('./extensionsManager')
const DocumentStore = require('./documentStore')
const createLogger = require('./logger')
const render = require('./render/render')
const safeSandbox = require('./render/safeSandbox')
const createNoneEngine = require('./render/noneEngine')
const htmlRecipe = require('./render/htmlRecipe')
const Reporter = require('../shared/reporter')

class WorkerReporter extends Reporter {
  #registry
  #initialized
  #documentStoreData
  #requestContextMetaConfigCollection

  constructor (workerData, registry) {
    const { options, documentStore, extensionsDefs } = workerData

    super(options)

    this.#registry = registry
    this.#initialized = false
    this.#documentStoreData = documentStore
    this.#requestContextMetaConfigCollection = new Map()

    this.requestModulesCache = new Map()

    this.beforeRenderListeners = this.createListenerCollection()
    this.afterTemplatingEnginesExecutedListeners = this.createListenerCollection()
    this.validateRenderListeners = this.createListenerCollection()

    this.logger = createLogger(registry)

    this.extensionsManager = ExtensionsManager(this, extensionsDefs)
  }

  async init () {
    if (this.#initialized === true) {
      throw new Error('jsreport already initialized. Make sure init is called only once')
    }

    super.init()

    await this.extensionsManager.init()

    this.documentStore = DocumentStore(this.#registry, this.#documentStoreData)

    this.addRequestContextMetaConfig('rootId', { sandboxReadOnly: true })
    this.addRequestContextMetaConfig('id', { sandboxReadOnly: true })
    this.addRequestContextMetaConfig('reportCounter', { sandboxReadOnly: true })
    this.addRequestContextMetaConfig('startTimestamp', { sandboxReadOnly: true })
    this.addRequestContextMetaConfig('logs', { sandboxReadOnly: true })
    this.addRequestContextMetaConfig('isChildRequest', { sandboxReadOnly: true })
    this.addRequestContextMetaConfig('originalInputDataIsEmpty', { sandboxReadOnly: true })
    this.addRequestContextMetaConfig('skipModificationDateUpdate', { sandboxHidden: true })

    const { compile: compileNone, execute: executeNone } = createNoneEngine()

    this.extensionsManager.engines.push({
      name: 'none',
      compile: compileNone,
      execute: executeNone
    })

    this.extensionsManager.recipes.push({
      name: 'html',
      execute: htmlRecipe
    })

    await this.initializeListeners.fire()

    this.#initialized = true
  }

  /**
   * @public
   */
  addRequestContextMetaConfig (property, options) {
    this.#requestContextMetaConfigCollection.set(property, options)
  }

  /**
   * @public
   */
  getRequestContextMetaConfig (property) {
    if (property === undefined) {
      const all = {}

      for (const [key, value] of this.#requestContextMetaConfigCollection.entries()) {
        all[key] = value
      }

      return all
    }

    return this.#requestContextMetaConfigCollection.get(property)
  }

  render (req, parentReq) {
    return render(this, req, parentReq)
  }

  async runInSandbox (fn, options = {}) {
    const { getContext, onCompile } = options
    let initialContext

    if (getContext != null) {
      initialContext = await getContext()
    } else {
      initialContext = {}
    }

    const runId = `__${nanoid()}__`

    initialContext[runId] = fn

    const {
      sandbox: sandboxContext,
      console: consoleSandbox,
      contextifyValue,
      decontextifyValue,
      restore,
      run
    } = safeSandbox(
      initialContext,
      {
        errorPrefix: options.errorPrefix,
        onLog: options.onLog,
        formatError: options.formatError,
        propertiesConfig: options.propertiesConfig,
        globalModules: options.globalModules || [],
        allowedModules: options.allowedModules || [],
        requirePaths: options.requirePaths || [],
        requireMap: options.requireMap,
        modulesCache: options.sandboxModulesCache
      }
    )

    if (onCompile) {
      const evalFn = (code) => {
        const executionCode = `
          ;(async () => {
            ${code}
          })();
        `

        return run(executionCode)
      }

      await onCompile({ context: sandboxContext, eval: evalFn })
    }

    const promise = run(`
      ;(async () => {
        const fn = this['${runId}']
        delete this['${runId}']
        return fn(this)
      })();
    `)

    const result = await promise

    return result
  }
}

module.exports = WorkerReporter
